### PR TITLE
[REF] hr_holidays: refactor accrual computation

### PR DESCRIPTION
**Pull request by naja628**
Refactor `_process_accrual_plans`, so that accrual happen only on the
period boundaries, and not partially on carryover or level change dates.

This should fix the following bugs
---
* Create an accrual plan:
    * Accrues days at the start of the period.
    * Carryover date 1 July.
    * Has 1 milestone:
        * Accrues 12 days yearly.
        * Starts immediately on allocation start date.
* Create an allocation using the above for your user that starts on
  2024/12/01
* in the timeoff dashboard use *balance at $date* to inspect
  what the accrued days will be at given dates.
        * any date in 2025 before july (eg 2025/01/01) -> 13-ish days
          available (desired)
        * any date in 2025 after july 1st (eg 2025/07/02) -> 7-ish days
          (wrong -> should still be 13)

---
* Create an accrual plan. Set the carryover date as July 1st.
* Create a milestone:
       * Set the number of accrued days to 10.
       * Set the accrual frequency as `Yearly` and the accrual date as January 1st.
       * Set the carryover policy `all time off carried over`
* Create an allocation. Set the start date as 01/01/2024. Set the
  accrual plan as the one created above.
* Use future allocations to see the number of allocated days on 01/01/2025. It should be 10.
* Use future allocations to see the number of allocated days on 01/07/2025. It should be still 10.
  Instead it will be 15. 5 days were accrued on the carryover date because half of the accrual period
  (half a year) has passed.

---
* Create an accrual plan:
        * Accrue at the start of the period
        * Milestone
                * accrue 12 days yearly
* Create an allocation with this plan, starting 1 month in the past
        * => 13 days will alloctaed
* Before confirming change the starting to 2 months in the past
        * => 12 days will be allocated (should be 14)

task-4115939

### CODE REVIEW

**Comment by naja628** [comment]
I changed a few test, but always with a rationale -- i.e. the new behavior is more in line what we ideally want. (Usually the old test's result came from a carryover being applied directly after an accrual)

File: addons/hr_holidays/models/hr_leave_allocation.py [hunk 1]
```diff
@@ -469,77 +470,134 @@ def _process_accrual_plans(self, date_to=False, force_period=False, log=True):
...
+                                current_level_maximum_leave = get_max_days(allocation, current_level)
+                                current_level_idx += 1
+                            # Last iteration
+                            allocation.with_context(
+                                override_accrual_dates=(subperiod_start, period_end)
+                            )._add_days_to_allocation(
+                                current_level, current_level_maximum_leave, leaves_taken, period_start, period_end
+                            )
+                        else:
+                            allocation._add_days_to_allocation(current_level, current_level_maximum_leave, leaves_taken, period_start, period_end)
```
Comment (Flotchet) [inline addons/hr_holidays/models/hr_leave_allocation.py:594]: At this point it start to get really hard to follow tbh maybe it could be worth to create 2 helper function for it or 

also I see that this part
```py
                    current_level, current_level_idx, current_level_maximum_leave = get_level_data(allocation, allocation.nextcall)

                    # get the period start and end for the **ongoing** period (as of `allocation.nextcall`), since we accrue days in advance
                    period_start, period_end = period_boundaries(current_level, allocation.nextcall)
                    nextcall = period_end

                    # If there is a carryover date coming, nextcall is set to it -- so the cron is called when we need to remove days
                    carryover_date = allocation._get_carryover_date(allocation.nextcall)
                    if allocation.nextcall < carryover_date < nextcall:
                        nextcall = min(nextcall, carryover_date)
                    # if it's the carry-over date, adjust days using current level's carry-over policy, then continue
                    if allocation.nextcall == carryover_date:
                        carryover(allocation, current_level)
```

is basically duplicated as well as 

```py
                     allocation.lastcall = allocation.nextcall
                    allocation.nextcall = nextcall
                    if force_period and allocation.nextcall > date_to:
                        allocation.nextcall = date_to
                        force_period = False
```
            
I'm wondering if creating a method of the structure

```py
                    current_level, current_level_idx, current_level_maximum_leave = get_level_data(allocation, allocation.nextcall)

                    # get the period start and end for the **ongoing** period (as of `allocation.nextcall`), since we accrue days in advance
                    period_start, period_end = period_boundaries(current_level, allocation.nextcall)
                    nextcall = period_end

                    # If there is a carryover date coming, nextcall is set to it -- so the cron is called when we need to remove days
                    carryover_date = allocation._get_carryover_date(allocation.nextcall)
                    if allocation.nextcall < carryover_date < nextcall:
                        nextcall = min(nextcall, carryover_date)
                    # if it's the carry-over date, adjust days using current level's carry-over policy, then continue
                    if allocation.nextcall == carryover_date:
                        carryover(allocation, current_level)

###### biffurcation call #######

                    allocation.lastcall = allocation.nextcall
                    allocation.nextcall = nextcall
                    if force_period and allocation.nextcall > date_to:
                        allocation.nextcall = date_to
                        force_period = False

```

would be better
it's an open question tho really not sure on this one I make it that way because of the number of args

**Review by Flotchet — CHANGES_REQUESTED** [review]
Hello thanks for brain damage! :brain: 
Actually this code is quite hard to follow so I prefer to work in multiple rounds on this one! :eyes:

File: addons/hr_holidays/models/hr_leave_allocation.py [hunk 2]
```diff
@@ -469,77 +470,134 @@ def _process_accrual_plans(self, date_to=False, force_period=False, log=True):
             first_level_start_date = allocation.date_from + get_timedelta(first_level.start_count, first_level.start_type)
             leaves_taken = allocation.leaves_taken if first_level.added_value_type == "day" else allocation.leaves_taken / (allocation.employee_id.sudo().resource_id.calendar_id.hours_per_day or HOURS_PER_DAY)
             # first time the plan is run, initialize nextcall and take carryover / level transition into account
+            at_period_start = (allocation.accrual_plan_id.accrued_gain_time == 'start')
```
[the remark this answers is not in the document]
Comment (naja628) [reply addons/hr_holidays/models/hr_leave_allocation.py:473]: Personally I think it reads better

**Review by Bertrand2 — CHANGES_REQUESTED** [review]
Sorry for the delay, the `_update_accrual_plans` method being what it is, it requires quite the time to wrap my head around it :sweat_smile:
If possible I'd say that this PR should try to stick as much as possible to being a fix and modify the least amount of code as possible; a refactor in stable seems like a stretch as we're limited to keep method signatures and some modifications could end up adding complexity. For a truly efficient refactor, it should be done in master.
Some points that you might need to keep an eye on:
- If extra methods/functions are needed, prioritize extra methods rather than functions inside methods as they could be useful in other places.
- Huge "if/else" are a risk as they might lead to code repetition; check if a similar logic could be applied or use more local conditions/variables
